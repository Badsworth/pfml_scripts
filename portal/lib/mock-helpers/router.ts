/**
 * @file Common next/router mocks for use in Storybook and Tests
 */
import Router from "next/router";
import createMockFn from "./createMockFn";

export interface MockRouterEvent {
  name: string;
  callback: RouterEventCallback;
}

export const mockRouterEvents: MockRouterEvent[] = [];

type RouterEventCallback = (...args: unknown[]) => unknown;

// @ts-expect-error Router mock is missing properties we don't currently reference
export const mockRouter = {
  locale: "en-US",
  route: "/",
  pathname: "",
  query: {},
  asPath: "",
  back: createMockFn(),
  beforePopState: createMockFn(),
  prefetch: createMockFn(() => Promise.resolve()),
  push: createMockFn(),
  replace: createMockFn(),
  reload: createMockFn(),
  events: {
    on: (name: string, callback: RouterEventCallback) => {
      mockRouterEvents.push({ name, callback });
    },
    off: createMockFn(),
  },
  isFallback: false,
} as typeof Router.router;

export function useRouter() {
  return mockRouter;
}
