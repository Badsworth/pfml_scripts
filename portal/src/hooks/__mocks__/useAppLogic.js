import Claim from "../../models/Claim";
import ClaimCollection from "../../models/ClaimCollection";
import { uniqueId } from "lodash";

export default jest.fn(() => ({
  auth: {
    createAccount: jest.fn(),
    forgotPassword: jest.fn(),
    login: jest.fn(),
  },
  claims: new ClaimCollection(),
  createClaim: jest.fn(() => new Claim({ application_id: uniqueId() })),
  loadClaims: jest.fn(),
  submitClaim: jest.fn(),
  updateClaim: jest.fn(),
}));
