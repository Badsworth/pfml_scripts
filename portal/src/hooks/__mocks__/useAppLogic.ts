import {
  BenefitsApplicationDocument,
  ClaimDocument,
} from "../../models/Document";
import User, { UserLeaveAdministrator } from "../../models/User";
import ApiResourceCollection from "../../models/ApiResourceCollection";
import BenefitsApplication from "../../models/BenefitsApplication";
import Claim from "../../models/Claim";
import ClaimDetail from "../../models/ClaimDetail";
import EmployerClaim from "../../models/EmployerClaim";
import Flag from "../../models/Flag";
import { Payment } from "src/models/Payment";
import { uniqueId } from "lodash";

export default jest.fn(() => ({
  errors: [],
  auth: {
    createAccount: jest.fn(),
    createEmployerAccount: jest.fn(),
    forgotPassword: jest.fn(),
    isLoggedIn: true,
    login: jest.fn(),
    logout: jest.fn(),
    requireLogin: jest.fn(),
    resendForgotPasswordCode: jest.fn(),
    resendVerifyAccountCode: jest.fn(),
    resetPassword: jest.fn(),
    verifyAccount: jest.fn(),
  },
  catchError: jest.fn(),
  benefitsApplications: {
    benefitsApplications: new ApiResourceCollection<BenefitsApplication>(
      "application_id"
    ),
    complete: jest.fn(),
    create: jest.fn(
      () => new BenefitsApplication({ application_id: uniqueId() })
    ),
    get: jest.fn(),
    hasLoadedBenefitsApplicationAndWarnings: jest.fn().mockReturnValue(true),
    hasLoadedAll: false,
    load: jest.fn(),
    loadAll: jest.fn(),
    submit: jest.fn(),
    submitPaymentPreference: jest.fn(),
    update: jest.fn(),
    warningsLists: {},
  },
  claims: {
    claimDetail: new ClaimDetail(),
    claims: new ApiResourceCollection<Claim>("fineos_absence_id"),
    clearClaims: jest.fn(),
    isLoadingClaims: null,
    isLoadingClaimDetail: null,
    loadClaimDetail: jest.fn(),
    paginationMeta: {},
    loadPage: jest.fn(),
  },
  clearErrors: jest.fn(),
  clearRequiredFieldErrors: jest.fn(),
  documents: {
    attach: jest.fn((application_id, files) => {
      const uploadPromises: Array<Promise<{ success: boolean }>> = [];
      for (let i = 0; i < files.length; i++) {
        uploadPromises.push(Promise.resolve({ success: true }));
      }
      return uploadPromises;
    }),
    hasLoadedClaimDocuments: jest.fn(),
    documents: new ApiResourceCollection<BenefitsApplicationDocument>(
      "fineos_document_id"
    ),
    download: jest.fn(),
    loadAll: jest.fn(),
  },
  employers: {
    addEmployer: jest.fn(),
    downloadDocument: jest.fn(() => new Blob()),
    loadClaim: jest.fn(
      () =>
        new EmployerClaim({
          absence_periods: [],
          fineos_absence_id: "NTN-111-ABS-01",
        })
    ),
    loadDocuments: jest.fn(
      () => new ApiResourceCollection<ClaimDocument>("fineos_document_id")
    ),
    loadWithholding: jest.fn(() => ({ filing_period: "2011-11-20" })),
    submitClaimReview: jest.fn(),
    submitWithholding: jest.fn(),
  },
  payments: {
    loadPayments: jest.fn(),
    loadedPaymentsData: new Payment(),
    hasLoadedPayments: true,
    isLoadingPayments: false,
  },
  portalFlow: {
    getNextPageRoute: jest.fn(),
    goTo: jest.fn(),
    goToNextPage: jest.fn(),
    goToPageFor: jest.fn(),
    pathname: "/mock-portalFlow-pathname",
  },
  otherLeaves: {
    removeEmployerBenefit: jest.fn(() => true),
    removeOtherIncome: jest.fn(() => true),
    removePreviousLeave: jest.fn(() => true),
  },
  setErrors: jest.fn(),
  updateUser: jest.fn(),
  user: new User({
    user_id: "mock_user_id",
    consented_to_data_sharing: true,
    email_address: "ali@miau.com",
    auth_id: "cognito_123",
    roles: [],
    user_leave_administrators: [],
  }),
  users: {
    loadUser: jest.fn(),
    requireUserConsentToDataAgreement: jest.fn(),
    requireUserRole: jest.fn(),
    updateUser: jest.fn(),
    user: new User({
      auth_id: "mock_auth_id",
      user_id: "mock_user_id",
      consented_to_data_sharing: true,
      user_leave_administrators: [
        // not verified, but can be
        new UserLeaveAdministrator({
          employer_dba: "Book Bindings 'R Us",
          employer_fein: "**-***1823",
          employer_id: "dda903f-f093f-ff900",
          has_fineos_registration: true,
          has_verification_data: true,
          verified: false,
        }),
        // already verified
        new UserLeaveAdministrator({
          employer_dba: "Knitting Castle",
          employer_fein: "**-***3443",
          employer_id: "dda930f-93jfk-iej08",
          has_fineos_registration: true,
          has_verification_data: true,
          verified: true,
        }),
        // not verified and cannot be verified
        new UserLeaveAdministrator({
          employer_dba: "Tomato Touchdown",
          employer_fein: "**-***7192",
          employer_id: "io19fj9-00jjf-uiw3r",
          has_fineos_registration: true,
          has_verification_data: false,
          verified: false,
        }),
      ],
    }),
  },
  featureFlags: {
    flags: [
      new Flag({
        enabled: false,
        name: "maintenance",
        options: {
          page_routes: ["/*"],
        },
        start: null,
        end: null,
      }),
    ],
    getFlag: jest.fn(() => {
      return new Flag({
        enabled: false,
        name: "maintenance",
        options: {
          page_routes: ["/*"],
        },
        start: null,
        end: null,
      });
    }),
    loadFlags: jest.fn(),
  },
}));
