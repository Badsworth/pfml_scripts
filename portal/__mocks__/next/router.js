import { mockRouter, mockRouterEvents } from "../../lib/mock-helpers/router";
export * from "../../lib/mock-helpers/router";

afterEach(() => {
  mockRouter.pathname = "";

  // Clear mock router events
  mockRouterEvents.length = 0;
});
