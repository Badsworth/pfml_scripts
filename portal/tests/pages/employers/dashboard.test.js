import Claim, { ClaimEmployee, ClaimEmployer } from "../../../src/models/Claim";
import User, { UserLeaveAdministrator } from "../../../src/models/User";
import { renderWithAppLogic, simulateEvents, testHook } from "../../test-utils";
import ClaimCollection from "../../../src/models/ClaimCollection";
import Dashboard from "../../../src/pages/employers/dashboard";
import PaginationMeta from "../../../src/models/PaginationMeta";
import faker from "faker";
import { mockRouter } from "next/router";
import routes from "../../../src/routes";
import useAppLogic from "../../../src/hooks/useAppLogic";

function createUserLeaveAdministrator(attrs = {}) {
  return new UserLeaveAdministrator({
    employer_id: faker.datatype.uuid(),
    employer_dba: faker.company.companyName(),
    employer_fein: `${faker.finance.account(2)}-${faker.finance.account(7)}`,
    ...attrs,
  });
}

const verifiedUserLeaveAdministrator = createUserLeaveAdministrator({
  employer_dba: "Work Inc",
  employer_fein: "12-3456789",
  has_fineos_registration: true,
  has_verification_data: true,
  verified: true,
});

const verifiableUserLeaveAdministrator = createUserLeaveAdministrator({
  employer_dba: "Book Bindings 'R Us",
  has_fineos_registration: false,
  has_verification_data: true,
  verified: false,
});

const getClaims = (leaveAdmin) => {
  return [
    new Claim({
      created_at: "2021-01-15",
      employee: new ClaimEmployee({
        first_name: "Jane",
        middle_name: null,
        last_name: "Doe",
      }),
      employer: new ClaimEmployer({
        employer_dba: leaveAdmin.employer_dba,
        employer_fein: leaveAdmin.employer_fein,
        employer_id: leaveAdmin.employer_id,
      }),
      fineos_absence_id: "NTN-111-ABS-01",
      claim_status: "Approved",
    }),
  ];
};

const setup = ({
  activeFilters = {},
  claims = [],
  userAttrs = {},
  paginationMeta = {},
  query = {},
} = {}) => {
  let appLogic;
  // Need to set an accurate pathname so portalFlow can return the correct links to route to
  mockRouter.pathname = routes.employers.dashboard;

  testHook(() => {
    appLogic = useAppLogic();
    // Fulfill the needs of withClaims to simulate that the user can view the page,
    // and that a list of claims has been loaded
    appLogic.users.user = new User({
      consented_to_data_sharing: true,
      ...userAttrs,
    });
    appLogic.claims.activeFilters = activeFilters;
    appLogic.claims.claims = new ClaimCollection(claims);
    appLogic.claims.shouldLoadPage = jest.fn().mockReturnValue(false);
    appLogic.claims.paginationMeta = new PaginationMeta({
      page_offset: 1,
      page_size: 25,
      total_pages: 3,
      total_records: 75,
      order_by: "created_at",
      order_direction: "asc",
      ...paginationMeta,
    });
  });
  const updateQuerySpy = jest.spyOn(appLogic.portalFlow, "updateQuery");

  const { wrapper } = renderWithAppLogic(Dashboard, {
    props: { appLogic, query },
    userAttrs,
  });

  return {
    appLogic,
    updateQuerySpy,
    wrapper,
  };
};

describe("Employer dashboard", () => {
  it("renders the page with expected content and pagination components", () => {
    const { wrapper } = setup();

    // Take targeted snapshots of content elements to avoid snapshotting noisy props
    expect(wrapper.find("Title")).toMatchSnapshot();
    wrapper.find("p").forEach((el) => expect(el).toMatchSnapshot());
    wrapper
      .find("Trans")
      .forEach((trans) => expect(trans.dive()).toMatchSnapshot());

    expect(wrapper.find("Details")).toMatchSnapshot();
    expect(wrapper.find("PaginationSummary")).toMatchSnapshot();
    expect(wrapper.find("PaginationNavigation")).toMatchSnapshot();
  });

  it("renders a beta info alert if all employers are registered in FINEOS", () => {
    const { wrapper } = setup({
      userAttrs: {
        user_leave_administrators: [
          new UserLeaveAdministrator({
            employer_dba: "Work Inc",
            employer_fein: "12-3456789",
            employer_id: "mock-employer-id-1",
            has_fineos_registration: true,
            has_verification_data: true,
            verified: true,
          }),
        ],
      },
    });

    expect(
      wrapper.find("DashboardInfoAlert").dive().find("Alert").prop("heading")
    ).toMatchInlineSnapshot(
      `"We're making it easier to manage paid leave applications"`
    );
    expect(
      wrapper.find("DashboardInfoAlert").dive().find("Alert Trans").dive()
    ).toMatchSnapshot();
  });

  it("renders an alert about claim availability when some employers aren't registered in FINEOS yet", () => {
    const { wrapper } = setup({
      userAttrs: {
        user_leave_administrators: [
          new UserLeaveAdministrator({
            employer_dba: "Work Inc",
            employer_fein: "12-3456789",
            employer_id: "mock-employer-id-1",
            has_fineos_registration: false,
            has_verification_data: true,
            verified: true,
          }),
          new UserLeaveAdministrator({
            employer_dba: "Work Co",
            employer_fein: "00-3456789",
            employer_id: "mock-employer-id-2",
            has_fineos_registration: false,
            has_verification_data: true,
            verified: true,
          }),
        ],
      },
    });

    expect(
      wrapper.find("DashboardInfoAlert").dive().find("Alert").prop("heading")
    ).toMatchInlineSnapshot(
      `"Your applications are not accessible right now for: 12-3456789, 00-3456789"`
    );

    expect(
      wrapper.find("DashboardInfoAlert").dive().find("Alert Trans").dive()
    ).toMatchSnapshot();
  });

  it("renders a table of claims with links if employer is registered in FINEOS", () => {
    const claims = getClaims(verifiedUserLeaveAdministrator);
    const userAttrs = {
      // Set multiple employers so the table shows all possible columns
      user_leave_administrators: [
        verifiedUserLeaveAdministrator,
        verifiableUserLeaveAdministrator,
      ],
    };

    const { wrapper } = setup({ claims, userAttrs });

    expect(wrapper.find("ClaimTableRows").dive()).toMatchSnapshot();
    expect(wrapper.find("thead")).toMatchSnapshot();
  });

  it("renders claim rows without links if employer is not registered in FINEOS", () => {
    const verifiedButNotInFineos = {
      ...verifiableUserLeaveAdministrator,
      verified: true,
    };
    const claims = getClaims(verifiedButNotInFineos);

    const userAttrs = {
      user_leave_administrators: [verifiedButNotInFineos],
    };

    const { wrapper } = setup({ claims, userAttrs });

    expect(wrapper.find("ClaimTableRows").dive().find("a")).toHaveLength(0);
  });

  it("allows Claim.employee to be null", () => {
    let claims = getClaims(verifiedUserLeaveAdministrator);
    claims = claims.map((claim) => {
      claim.employee = null;
      return claim;
    });

    const { wrapper } = setup({
      claims,
      userAttrs: {
        user_leave_administrators: [verifiedUserLeaveAdministrator],
      },
    });

    expect(
      wrapper
        .find("ClaimTableRows")
        .dive()
        .find('[data-test="employee_name"] a')
        .text()
    ).toBe("--");
  });

  it("does not render Employer DBA when user has only one Employer associated", () => {
    const { wrapper: soloEmployerWrapper } = setup({
      userAttrs: {
        user_leave_administrators: [verifiedUserLeaveAdministrator],
      },
    });
    const { wrapper: multipleEmployerWrapper } = setup({
      userAttrs: {
        user_leave_administrators: [
          verifiedUserLeaveAdministrator,
          verifiableUserLeaveAdministrator,
        ],
      },
    });

    expect(soloEmployerWrapper.find("ClaimTableRows").prop("tableColumnKeys"))
      .toMatchInlineSnapshot(`
      Array [
        "employee_name",
        "fineos_absence_id",
        "employer_fein",
        "created_at",
        "status",
      ]
    `);
    expect(
      multipleEmployerWrapper.find("ClaimTableRows").prop("tableColumnKeys")
    ).toMatchInlineSnapshot(`
      Array [
        "employee_name",
        "fineos_absence_id",
        "employer_dba",
        "employer_fein",
        "created_at",
        "status",
      ]
    `);
  });

  it("renders a 'no results' message in the table, and no pagination components when no claims are present", () => {
    const { wrapper } = setup({
      userAttrs: {
        user_leave_administrators: [verifiedUserLeaveAdministrator],
      },
      paginationMeta: {
        total_records: 0,
        total_pages: 1,
      },
    });

    expect(wrapper.find("ClaimTableRows").dive()).toMatchSnapshot();
    expect(wrapper.find("PaginationSummary").exists()).toBe(false);
    expect(wrapper.find("PaginationNavigation").exists()).toBe(false);
  });

  it("renders only the pagination summary when only one page of claims exists", () => {
    const { wrapper } = setup({
      paginationMeta: {
        total_records: 25,
        total_pages: 1,
      },
    });

    expect(wrapper.find("PaginationSummary").exists()).toBe(true);
    expect(wrapper.find("PaginationNavigation").exists()).toBe(false);
  });

  it("changes the page_offset query param when a page navigation button is clicked", () => {
    const { updateQuerySpy, wrapper } = setup();
    const clickedPageOffset = "3";

    wrapper.find("PaginationNavigation").simulate("click", clickedPageOffset);

    expect(updateQuerySpy).toHaveBeenCalledWith({
      page_offset: clickedPageOffset,
    });
  });

  it("renders the banner if there are any unverified employers", () => {
    const { wrapper } = setup({
      userAttrs: {
        user_leave_administrators: [
          // Mix of verified and unverified
          verifiedUserLeaveAdministrator,
          verifiableUserLeaveAdministrator,
        ],
      },
    });
    expect(wrapper.find("Alert").exists()).toEqual(true);
  });

  it("renders instructions if there are no verified employers", () => {
    const { wrapper } = setup({
      userAttrs: {
        // No verified employers
        user_leave_administrators: [verifiableUserLeaveAdministrator],
      },
    });

    expect(
      wrapper.find("[data-test='verification-instructions-row'] Trans").dive()
    ).toMatchSnapshot();
  });

  it("does not render search section when no search feature flag is enabled", () => {
    const { wrapper } = setup();

    expect(wrapper.find("Search").dive().isEmptyRender()).toBe(true);
  });

  it("renders search section when feature flags are enabled", () => {
    process.env.featureFlags = {
      employerShowDashboardSearch: true,
    };

    const { wrapper } = setup({
      activeFilters: {
        search: "Initial search field value",
      },
    });

    expect(wrapper.find("Search").dive()).toMatchSnapshot();
  });

  it("updates search param when a search is performed", async () => {
    process.env.featureFlags = {
      employerShowDashboardSearch: true,
    };

    const { updateQuerySpy, wrapper } = setup();

    const search = wrapper.find("Search").dive();
    const { changeField, submitForm } = simulateEvents(search);

    changeField("search", "Bud Baxter");
    await submitForm();

    expect(updateQuerySpy).toHaveBeenCalledWith({
      page_offset: "1",
      search: "Bud Baxter",
    });
  });

  it("does not render filters when no filter feature flags are enabled", () => {
    const { wrapper } = setup();

    expect(wrapper.find("Filters").dive().isEmptyRender()).toBe(true);
  });

  it("renders filters when feature flags are enabled", () => {
    process.env.featureFlags = {
      employerShowDashboardFilters: true,
    };

    const { wrapper } = setup({
      userAttrs: {
        // Include multiple LA's so Employer filter shows
        user_leave_administrators: [
          createUserLeaveAdministrator({
            employer_dba: "Work Inc",
            employer_id: "mock-id-1",
            employer_fein: "11-3456789",
            verified: true,
          }),
          createUserLeaveAdministrator({
            employer_dba: "Acme",
            employer_id: "mock-id-2",
            employer_fein: "22-3456789",
            verified: true,
          }),
        ],
      },
    });

    expect(wrapper.find("Filters").dive()).toMatchSnapshot();
  });

  it("renders filters toggle button with expected label and aria attributes", () => {
    process.env.featureFlags = {
      employerShowDashboardFilters: true,
    };

    const user_leave_administrators = [
      createUserLeaveAdministrator({
        verified: true,
      }),
    ];
    const { wrapper: CollapsedWithoutFilters } = setup();
    const { wrapper: CollapsedWithFilters } = setup({
      activeFilters: {
        claim_status: "Pending",
        employer_id: user_leave_administrators[0].employer_id,
      },
      userAttrs: {
        // Include multiple LA's so Employer filter shows
        user_leave_administrators,
      },
    });
    const { wrapper: ExpandedWithoutFilters } = setup({
      query: {
        "show-filters": "true",
      },
    });

    const getButton = (w) =>
      w.find("Filters").dive().find("Button[aria-controls='filters']");

    expect(getButton(CollapsedWithFilters)).toMatchSnapshot();
    expect(getButton(CollapsedWithoutFilters)).toMatchSnapshot();
    expect(getButton(ExpandedWithoutFilters)).toMatchSnapshot();
  });

  it("sets initial filter form state from activeFilters prop", () => {
    process.env.featureFlags = {
      employerShowDashboardFilters: true,
    };
    // Include multiple LA's so Employer filter shows
    const user_leave_administrators = [
      createUserLeaveAdministrator({
        verified: true,
      }),
      createUserLeaveAdministrator({
        verified: true,
      }),
    ];

    const { wrapper } = setup({
      activeFilters: {
        claim_status: "Approved,Closed",
        employer_id: user_leave_administrators[0].employer_id,
      },
      userAttrs: {
        user_leave_administrators,
      },
    });

    const filters = wrapper.find("Filters").dive();

    expect(filters.find("Dropdown").prop("value")).toBe(
      user_leave_administrators[0].employer_id
    );
    expect(
      filters
        .find("InputChoiceGroup")
        .prop("choices")
        .map((choice) => choice.checked)
    ).toMatchInlineSnapshot(`
      Array [
        true,
        true,
        false,
        false,
      ]
    `);
  });

  it("shows filters section when show-filters query param is set", () => {
    process.env.featureFlags = {
      employerShowDashboardFilters: true,
    };

    const { wrapper: collapsedWrapper } = setup();
    const { wrapper: expandedWrapper } = setup({
      query: {
        "show-filters": "true",
      },
    });

    const getFiltersContainer = (w) =>
      w.find("Filters").dive().find("#filters");

    expect(getFiltersContainer(collapsedWrapper).prop("hidden")).toBe(true);
    expect(getFiltersContainer(expandedWrapper).prop("hidden")).toBe(false);
  });

  it("toggles show-filters param when toggle button is clicked", () => {
    process.env.featureFlags = {
      employerShowDashboardFilters: true,
    };

    const { updateQuerySpy, wrapper } = setup();
    const filters = wrapper.find("Filters").dive();
    const { click } = simulateEvents(filters);

    click("Button[aria-controls='filters']");

    expect(updateQuerySpy).toHaveBeenCalledWith({ "show-filters": "true" });
  });

  it("renders organizations filter when there are multiple verified organizations", () => {
    process.env.featureFlags = { employerShowDashboardFilters: true };

    const { wrapper: wrapperWithOneVerifiedOrg } = setup({
      userAttrs: {
        user_leave_administrators: [
          createUserLeaveAdministrator({
            verified: false,
          }),
          createUserLeaveAdministrator({
            verified: true,
          }),
        ],
      },
    });

    const { wrapper: wrapperWithMultipleVerifiedOrgs } = setup({
      userAttrs: {
        user_leave_administrators: [
          createUserLeaveAdministrator({
            has_fineos_registration: false, // this employer should still show in the list
            verified: true,
          }),
          createUserLeaveAdministrator({
            has_fineos_registration: true,
            verified: true,
          }),
        ],
      },
    });

    const getEmployerFilterDropdown = (wrapper) =>
      wrapper.find("Filters").dive().find("Dropdown[name='employer_id']");

    expect(getEmployerFilterDropdown(wrapperWithOneVerifiedOrg).exists()).toBe(
      false
    );
    expect(
      getEmployerFilterDropdown(wrapperWithMultipleVerifiedOrgs).exists()
    ).toBe(true);
  });

  it("updates query params when user changes filter to Approved and Closed", async () => {
    expect.assertions();

    process.env.featureFlags = {
      employerShowDashboardFilters: true,
    };

    const user_leave_administrators = [
      createUserLeaveAdministrator({
        verified: true,
      }),
      createUserLeaveAdministrator({
        verified: true,
      }),
    ];

    const { updateQuerySpy, wrapper } = setup({
      userAttrs: {
        user_leave_administrators,
      },
      paginationMeta: {
        page_offset: 2,
      },
    });

    const filtersWrapper = wrapper.find("Filters").dive();
    const { changeField, changeCheckbox, submitForm } = simulateEvents(
      filtersWrapper
    );

    changeCheckbox("claim_status", true, "Approved");
    changeCheckbox("claim_status", true, "Closed");
    changeField("employer_id", user_leave_administrators[0].employer_id);

    expect(filtersWrapper.find("Button[type='submit']").prop("disabled")).toBe(
      false
    );

    await submitForm();
    expect(updateQuerySpy).toHaveBeenCalledWith({
      claim_status: "Approved,Closed",
      employer_id: user_leave_administrators[0].employer_id,
      page_offset: "1",
    });
  });

  it("resets query params when user clicks Reset action", () => {
    process.env.featureFlags = {
      employerShowDashboardFilters: true,
    };

    const user_leave_administrators = [
      createUserLeaveAdministrator({
        verified: true,
      }),
      createUserLeaveAdministrator({
        verified: true,
      }),
    ];

    const { updateQuerySpy, wrapper } = setup({
      activeFilters: {
        claim_status: "Approved",
        employer_id: user_leave_administrators[0].employer_id,
      },
      userAttrs: {
        user_leave_administrators,
      },
      paginationMeta: {
        page_offset: 2,
      },
    });

    const filtersWrapper = wrapper.find("Filters").dive();
    const { click } = simulateEvents(filtersWrapper);

    click("Button[data-test='reset-filters']");

    expect(updateQuerySpy).toHaveBeenCalledWith({
      page_offset: "1",
    });
  });

  it("shows buttons for removing active filters", () => {
    const user_leave_administrators = [
      createUserLeaveAdministrator({
        verified: true,
        employer_dba: "Acme Co",
      }),
    ];

    const { wrapper: wrapperWithActiveFilters } = setup({
      activeFilters: {
        claim_status: "Approved,Closed",
        employer_id: user_leave_administrators[0].employer_id,
      },
      userAttrs: {
        user_leave_administrators,
      },
    });
    const { wrapper: wrapperWithoutActiveFilters } = setup({
      userAttrs: {
        user_leave_administrators,
      },
    });

    const findFiltersMenu = (wrapper) =>
      wrapper.find("Filters").dive().find("[data-test='filters-menu']");

    expect(findFiltersMenu(wrapperWithoutActiveFilters).exists()).toBe(false);
    expect(findFiltersMenu(wrapperWithActiveFilters)).toMatchSnapshot();
  });

  it("removes filter when a filter menu button is clicked", () => {
    const user_leave_administrators = [
      createUserLeaveAdministrator({
        verified: true,
        employer_dba: "Acme Co",
      }),
    ];

    const { updateQuerySpy, wrapper } = setup({
      activeFilters: {
        claim_status: "Approved,Closed",
        employer_id: user_leave_administrators[0].employer_id,
      },
      userAttrs: {
        user_leave_administrators,
      },
    });
    const { click } = simulateEvents(wrapper.find("Filters").dive());

    click("FilterMenuButton[data-test='employer_id']");
    expect(updateQuerySpy).toHaveBeenLastCalledWith({
      page_offset: "1",
    });

    click("FilterMenuButton[data-test='claim_status_Closed']");
    expect(updateQuerySpy).toHaveBeenLastCalledWith({
      claim_status: "Approved",
      page_offset: "1",
    });
  });
});
