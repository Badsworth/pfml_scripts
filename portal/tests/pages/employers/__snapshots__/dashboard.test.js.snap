// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employer dashboard renders a 'no results' message in the table, and no pagination components when no claims are present 1`] = `
<tr>
  <td
    colSpan={5}
  >
    No applications on file
  </td>
</tr>
`;

exports[`Employer dashboard renders a beta info alert if all employers are registered in FINEOS 2`] = `
Array [
  "Our dashboard is in beta. You can now see all the applications you need to review using our new dashboard. Over the next few months, we'll be adding more features to the dashboard so you can easily find and manage paid leave applications. You can ",
  <a
    href="https://www.mass.gov/paidleave-employer-feedback"
    key="1"
    rel="noopener"
    target="_blank"
  >
    use our feedback form
  </a>,
  " to tell us what you think of our new features.",
]
`;

exports[`Employer dashboard renders a table of claims with links if employer is registered in FINEOS 1`] = `
<tr
  key="NTN-111-ABS-01"
>
  <th
    data-label="Employee name"
    data-test="employee_name"
    scope="row"
  >
    <Link
      href="/employers/applications/new-application?absence_id=NTN-111-ABS-01"
    >
      <a>
        Jane Doe
      </a>
    </Link>
  </th>
  <td
    data-label="Application ID"
    key="fineos_absence_id"
  >
    <Link
      href="/employers/applications/new-application?absence_id=NTN-111-ABS-01"
    >
      <a>
        NTN-111-ABS-01
      </a>
    </Link>
  </td>
  <td
    data-label="Organization"
    key="employer_dba"
  >
    Work Inc
  </td>
  <td
    data-label="Employer ID number"
    key="employer_fein"
  >
    12-3456789
  </td>
  <td
    data-label="Application start date"
    key="created_at"
  >
    1/15/2021
  </td>
  <td
    data-label="Status"
    key="status"
  >
    <AbsenceCaseStatusTag
      status="Approved"
    />
  </td>
</tr>
`;

exports[`Employer dashboard renders a table of claims with links if employer is registered in FINEOS 2`] = `
<thead>
  <tr>
    <th
      key="employee_name"
      scope="col"
    >
      Employee name
    </th>
    <th
      key="fineos_absence_id"
      scope="col"
    >
      Application ID
    </th>
    <th
      key="employer_dba"
      scope="col"
    >
      Organization
    </th>
    <th
      key="employer_fein"
      scope="col"
    >
      Employer ID number
    </th>
    <th
      key="created_at"
      scope="col"
    >
      Application start date
      <TooltipIcon
        position="bottom"
      >
        When an employee started a new leave application
      </TooltipIcon>
    </th>
    <th
      key="status"
      scope="col"
    >
      Status
    </th>
  </tr>
</thead>
`;

exports[`Employer dashboard renders an alert about claim availability when some employers aren't registered in FINEOS yet 2`] = `
Array [
  "Your account has been verified. It may take up to 15 minutes for our systems to update so that you can log in and review applications. If anyone else on your team needs to review applications, they’ll also need to complete the ",
  <a
    href="https://www.mass.gov/how-to/creating-an-employer-account-to-review-paid-family-and-medical-leave-pfml-applications"
    key="1"
    rel="noopener"
    target="_blank"
  >
    verification process
  </a>,
  ".",
]
`;

exports[`Employer dashboard renders claim rows without links if employer is not registered in FINEOS 1`] = `
<tr
  key="NTN-111-ABS-01"
>
  <th
    data-label="Employee name"
    data-test="employee_name"
    scope="row"
  >
    Jane Doe
  </th>
  <td
    data-label="Application ID"
    key="fineos_absence_id"
  >
    NTN-111-ABS-01
  </td>
  <td
    data-label="Employer ID number"
    key="employer_fein"
  >
    **-***0002
  </td>
  <td
    data-label="Application start date"
    key="created_at"
  >
    1/15/2021
  </td>
  <td
    data-label="Status"
    key="status"
  >
    <AbsenceCaseStatusTag
      status="Approved"
    />
  </td>
</tr>
`;

exports[`Employer dashboard renders the page with expected content and pagination components 1`] = `
<Title>
  Dashboard
</Title>
`;

exports[`Employer dashboard renders the page with expected content and pagination components 2`] = `
<p
  className="margin-y-2"
>
  Applications will not have a status until the Department has made a decision. Applications that don't have a status may require action from you.
</p>
`;

exports[`Employer dashboard renders the page with expected content and pagination components 3`] = `
Array [
  <strong
    key="strong-0"
  >
    “--” or no status:
  </strong>,
  " This leave request may require action from you; otherwise, it is awaiting a decision from DFML.",
]
`;

exports[`Employer dashboard renders the page with expected content and pagination components 4`] = `
Array [
  <strong
    key="strong-0"
  >
    Approved:
  </strong>,
  " DFML has approved this leave request.",
]
`;

exports[`Employer dashboard renders the page with expected content and pagination components 5`] = `
Array [
  <strong
    key="strong-0"
  >
    Closed:
  </strong>,
  " The leave has passed or no action can be taken.",
]
`;

exports[`Employer dashboard renders the page with expected content and pagination components 6`] = `
Array [
  <strong
    key="strong-0"
  >
    Denied:
  </strong>,
  " DFML has denied this leave request.",
]
`;

exports[`Employer dashboard renders the page with expected content and pagination components 7`] = `
<Details
  label="Status descriptions"
>
  <ul
    className="usa-list"
  >
    <li>
      <Trans
        i18nKey="pages.employersDashboard.statusDescription_none"
      />
    </li>
    <li>
      <Trans
        i18nKey="pages.employersDashboard.statusDescription_approved"
      />
    </li>
    <li>
      <Trans
        i18nKey="pages.employersDashboard.statusDescription_closed"
      />
    </li>
    <li>
      <Trans
        i18nKey="pages.employersDashboard.statusDescription_denied"
      />
    </li>
  </ul>
</Details>
`;

exports[`Employer dashboard renders the page with expected content and pagination components 8`] = `
<PaginationSummary
  pageOffset={1}
  pageSize={25}
  totalRecords={75}
/>
`;

exports[`Employer dashboard renders the page with expected content and pagination components 9`] = `
<PaginationNavigation
  onClick={[Function]}
  pageOffset={1}
  totalPages={3}
/>
`;

exports[`Employer dashboard when employerShowVerifications flag is enabled renders instructions if there are no verified employers 1`] = `
Array [
  "You have not verified any organizations. ",
  <a
    href="/employers/organizations"
    key="1"
  >
    Complete this process
  </a>,
  " to review applications.",
]
`;
